src/terrain/terrain_base.cpp

```diff
@@ -0,0 +1,223 @@
+#include "terrain_base.hpp"
+
+#include "../logging.hpp"
+#include "terrain_generation/noise.hpp"
+
+// move add_to_top
+
+namespace terrain {
+
+void
+TerrainBase::qb_read(
+    std::vector<uint32_t> data,
+    std::map<uint32_t, std::pair<const Material*, uint8_t>> materials_inverse
+) {
+    tiles_.reserve(X_MAX * Y_MAX * Z_MAX);
+
+    std::set<uint32_t> unknown_materials;
+
+    for (uint32_t xyz = 0; xyz < X_MAX * Y_MAX * Z_MAX; xyz++) {
+        auto [x, y, z] = sop(xyz);
+        uint32_t color = data[xyz];
+        if (color == 0) {                             // if the qb voxel is transparent.
+            auto mat_color = materials_inverse.at(0); // set the materials to air
+            tiles_.push_back(Tile({x, y, z}, mat_color.first, mat_color.second));
+        } else if (materials_inverse.count(color)) { // if the color is known
+            auto mat_color = materials_inverse.at(color);
+            tiles_.push_back(Tile({x, y, z}, mat_color.first, mat_color.second));
+        } else { // the color is unknown
+            unknown_materials.insert(color);
+            auto mat_color = materials_inverse.at(0); // else set to air.
+            tiles_.push_back(Tile({x, y, z}, mat_color.first, mat_color.second));
+        }
+    }
+
+    for (uint32_t color : unknown_materials) {
+        // is there any way to log hexadecimal
+        LOG_WARNING(logging::terrain_logger, "Cannot find color: {:x}", color);
+    }
+}
+
+int
+TerrainBase::get_first_not(
+    const std::set<std::pair<int, int>> materials, int x, int y, int guess
+) const {
+    if (guess < 1) {
+        guess = 1;
+    }
+    if ((uint32_t)guess >= Z_MAX) {
+        guess = Z_MAX - 1;
+    }
+    if (materials.find(std::make_pair(
+            get_tile(x, y, guess - 1)->get_material_id(),
+            get_tile(x, y, guess - 1)->get_color_id()
+        ))
+        != materials.end()) {
+        if (materials.find(std::make_pair(
+                get_tile(x, y, guess)->get_material_id(),
+                get_tile(x, y, guess)->get_color_id()
+            ))
+            == materials.end()) {
+            return guess;
+        } else {
+            // go up
+            for (uint16_t z = guess + 1; z < Z_MAX; z++) {
+                if (materials.find(std::make_pair(
+                        get_tile(x, y, z)->get_material_id(),
+                        get_tile(x, y, z)->get_color_id()
+                    ))
+                    == materials.end()) {
```

Comment:
This is horrendous extract the subexpression to a variable.

Also add some newlines